features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,373 @@
+  currentMessages = skygearChat.merge(currentMessages, messages);
+  // display currentMessages
+};
+
+skygearChat
+  .fetchMessages(conversation, beforeMessage, limit, order, successCallback)
+  .then(successCallback, errorCallback);
+
+// no-cache version
+skygearChat
+  .fetchMessages(conversation, beforeMessage, limit, order)
+  .then(successCallback, errorCallback);
+```
+
+#### Non-JS
+
+```
+skygearChat.fetchConversations(func (conversations, cached = false) {
+  // display conversations
+})
+```
+
+```
+let currentMessages = [];
+
+skygearChat.fetchMessages(func (messages, cached = false) {
+  currentMessages = skygearChat.merge(currentMessages, messages);
+  // display currentMessages
+})
+```
+
+- Add utility functions for handling result returned by chat container
+
+```
+func merge(conversationDeltas: [ConversationDelta], toList: [Conversation]) -> [Conversation]
+func merge(messages: [Message], toList: [Message]) -> [Message]
+func merge(messageDeltas: [MessageDelta], toList: [Message]) -> [Message]
+
+func findBoundaries(ofMessageList: [Message]) -> [(Int, Int)]
+```
+
+- Add a managed message list for user to use directly
+
+The ManagedMessageList is a default implementation provided by Skygear, its underlying implementation is just using the above new API and utility functions.
+
+```
+class ManagedMessageList {
+
+  constructor(conversation: Conversation,
+              delegate: ManagedMessageListDelegate)
+
+  - fetchLimit: Int = 100
+  - useCache: Boolean = true
+  - subscribeNewMessages: Boolean = true
+  - syncOnReconnect: Boolean = true
+
+  func online() // expected to call when enter view and pubsub connected
+
+  func offline() // expected to call when leave view and pubsub disconnected
+
+  func messageCount() -> Int
+
+  func messageAt(position: Int) -> Message
+
+  /**
+   * scrollViewDelegateMethods
+   *
+   * the list will determine when to fetch messages by checking current first
+   * and last visible cell
+   */
+
+}
+
+interface ManagedMessageListDelegate {
+
+  func managedMessageListDidChange(messageDeltas: [MessageDelta])
+
+}
+```
+
+# Example
+
+*All below UI update, I use `this.tableView.reloadTable()`. It is up to client app to decide how to update the UI*
+
+### In conversation list
+
+#### Start subscribing conversation changes
+
+When
+- enter conversation list
+- come offline and online again when in conversation list
+
+```
+skygearChat.subscribeConversations(completion: func (conversationDeltas) {
+  this.conversations = skygearChat.applyDelta(conversationDeltas, toList: this.conversations)
+  this.tableView.reloadTable()
+})
+```
+
+#### Fetch and update table
+
+When
+- enter conversation list
+- come offline and online again when in conversation list
+
+```
+skygearChat.fetchConversations(completion: func (conversations, cached = false) {
+  this.conversations = conversations
+  this.tableView.reloadTable()
+})
+```
+
+### In conversation OR app come back online inside conversation
```

Comment:
Similar to the question above, what it meant by "come back online inside conversation"?